### DISCUSSION TITLE
Support WHERE on querybuilder insert as part of conflict resolution

### DISCUSSION THREAD

**Discussion by noahw3** [mikro-orm/mikro-orm#4597] [Ideas]
**Is your feature request related to a problem? Please describe.**
I'm attempting to insert into a table using querybuilder. This table has a partial unique index. In order to correctly do this, Postgres needs the partial index to be specified in a WHERE clause. See https://stackoverflow.com/a/54169587 for more details.

However, it appears that when adding a WHERE clause to an insert query, it's entirely ignored.

**Describe the solution you'd like**
When specifying a WHERE condition using an insert query builder, it will be appended to the ON CONFLICT clause (at least in Postgres).

**Describe alternatives you've considered**
Arguably a better solution (though not mutually exclusive) would be to provide ON CONFLICT resolution as part of entity manager's create options. If I could just issue a create command and specify that I want conflicts to be ignored, that would be even better.

**Comment by noahw3** [comment]
After digging into the code, it appears that WHERE conditions are [already in use](https://github.com/mikro-orm/mikro-orm/blob/bd0724da417e7955c8ed982bcce1ffa68953c024/packages/knex/src/query/QueryBuilder.ts#L297) as part of the on conflict, but are only used as merge resolution and not as part of the actual conflict clause itself.

**Comment by B4nan** [comment]
This feature is just mapping how it works in knex, and it does not ignore the where clause - but you need to put it after the `onConflict` call. Please provide what you have tried, why you think it's not enough, and what exact query you want from it as a result. And maybe check the knex docs as well (and possibly issues) to understand how it works there, as here it's really the same thing. I see [they mention](https://knexjs.org/guide/query-builder.html#onconflict) you might need a raw statement for the partial index support: 

> For PostgreSQL and SQLite, you can use knex.raw(...) function in onConflict. It can be useful to specify condition when you have partial index :

But if they don't support it via the `where()` call, there must be some gotcha.

> Arguably a better solution (though not mutually exclusive) would be to provide ON CONFLICT resolution as part of entity manager's create options

You can use `em.upsert` for that, things are not that easy internally when it comes to state management and this methods tries to abstract it away. Under the hood it uses the on conflict queries and tries to respect what is already in the identity map. I can't see how a method like `em.create` could alter this, the method serves only one purpose - create entity instance, it does not touch the database on its own.

edit: maybe [this](https://github.com/knex/knex/issues/4590) is related? tbh I am starting to hate how everything a bit more complex in knex ends up as "use raw", for a library author this basically means "abstract everything better and do it yourself", which is indeed very helpful. I think you might be able to pass the raw instance from outside here as well, similarly to the knex docs, the `qb.raw()` will give you the raw knex QB instance.

**Reply by noahw3** [reply]
Sorry just getting back to this now. I played around with it a bit and I'm not sure how I'd be able to pass the raw knex string through, unless I just used the entire knex query builder I guess. Mikro's `onConflict` assumes that the parameters passed are `Fields` and tries to map them as such. https://github.com/mikro-orm/mikro-orm/blob/bd0724da417e7955c8ed982bcce1ffa68953c024/packages/knex/src/query/QueryBuilder.ts#L355

As you said, even if I could get this to work it would be quite clunky and unintuitive. Would it make sense to support it as a first class feature in mikro, and then abstract away the details of setting the `raw` string under the hood? Being a second optional parameter to `onConflict` seems ideal.

**Reply by B4nan** [reply]
Can you give me an example code of how you would like to use it? I need something concrete to understand better what you'd like to do and how the query should look like. 

edit: I'd like to make the `em.upsert` calls also more configurable, so feel free to propose how it could look with that too.

**Reply by noahw3** [reply]
So for a unique partial index like `create unique index single_self_assignment_index on bundle_assignments (bundle_id, assigner_id, assignment_type) where assignment_type = 'self';`

I'm thinking something along the lines of `onConflict(["bundle_id", "assignment_type", "assigner_id"], { assignmentType:  ASSIGNMENT_TYPES.SELF })`? Essentially including an entire WHERE condition as the second parameter to append as part of the `onConflict` clause.

The other obvious alternative is to add a new query builder method specifically for this purpose (something like `onConflictWhere`), but that seems worse since it pollutes the available methods when it _always_ has to be tied to an `onConflict` clause anyway.

As a side note, I'm not sure if the behavior is the same elsewhere in queryBuilder but I find it odd that the keys to the `onConflict` condition don't get mapped to the naming strategy. `Field<T>` is typed specifically as `keyof T`, but that actually does not work as they get passed straight through.
